oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelineMongoDownloadTask.java

```diff
@@ -0,0 +1,278 @@
+            this.lastIdDownloaded = null;
+            Instant failureTimestamp = null;
+            long retryInterval = retryInitialIntervalMillis;
+
+            Map<String, Integer> exceptions = new HashMap<>();
+            while (true) {
+                try {
+                    if (lastIdDownloaded != null) {
+                        LOG.info("Recovering from broken connection, finishing downloading documents with _modified={}", nextLastModified);
+                        downloadRange(new TraversingRange(new LastModifiedRange(nextLastModified, nextLastModified + 1), lastIdDownloaded));
+                        // Reset the retries
+                        failureTimestamp = null;
+                        // Continue downloading everything starting from the next _lastmodified value
+                        downloadRange(new TraversingRange(new LastModifiedRange(nextLastModified + 1, upperBoundLastModified), null));
+                    } else {
+                        downloadRange(new TraversingRange(new LastModifiedRange(nextLastModified, upperBoundLastModified), null));
+                    }
+                    LOG.info("Terminating thread, processed {} documents", documentsRead);
+                    return new Result(documentsRead);
+                } catch (MongoException e) {
+                    if (failureTimestamp == null) {
+                        failureTimestamp = Instant.now();
+                    }
+                    LOG.warn("Connection error downloading from MongoDB.", e);
+                    if (Duration.between(failureTimestamp, Instant.now()).toSeconds() > retryDuringSeconds) {
+                        // Give up. Get a string of all exceptions that were thrown
+                        StringBuilder summary = new StringBuilder();
+                        for (Map.Entry<String, Integer> entry : exceptions.entrySet()) {
+                            summary.append("\n\t").append(entry.getValue()).append("x: ").append(entry.getKey());
+                        }
+                        throw new RetryException(retryDuringSeconds, summary.toString(), e);
+                    } else {
+                        LOG.warn("Retrying download after {} ms; number of times failed: {}", retryInterval, failureTimestamp);
+                        exceptions.compute(e.getClass().getSimpleName() + " - " + e.getMessage(),
+                                (key, val) -> val == null ? 1 : val + 1
+                        );
+                        try {
+                            Thread.sleep(retryInterval);
+                        } catch (InterruptedException ignore) {
+                        }
+                        // simple exponential backoff mechanism
+                        retryInterval = Math.min(retryMaxIntervalMillis, retryInterval * 2);
+                    }
+                }
+            }
+        } catch (Throwable t) {
+            LOG.warn("Thread terminating with exception: " + t);
```

Comment:
I'm trying to avoid polluting the logs with multiple repetitions of the same stack trace. The exception here is propagated and caught by the main thread that is waiting for these tasks to complete (In the PipelinedStrategy class). And then the main thread rethrows the exception wrapped in a RuntimeException. The code that called the `createdSortedStoreFile()` will catch and finally log the exception with the stack trace.

Actually, there is a corner case where the exception will not be logged completely, which is if the main thread already stopped waiting for some of these futures, for instance, if one of the tasks aborted which will cause the main thread to abort with the exception of the first task that failed. In this case, we would only get the name and message of the exceptions thrown by the futures that failed later. Ok, that stack trace may be useful, so I will log the stack trace here as well (except for InterruptedExceptions).